go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+# Go SDK Workflow Interceptor Updates and OpenTelemetry Tracing Support
+
+This proposal is for updates to Go workflow interceptors with the following features:
+
+* Support for and implementation of OpenTelemetry tracing
+* Support idiomatic Go interceptor design
+* Provide feature parity with Java SDK and other missing features (see https://github.com/temporalio/sdk-go/issues/529)
+* Add activity interceptors
+* Make inbound interceptors future-proof for argument addition
+
+**These updates are known to NOT be backwards-compatible**
+
+## Interceptor Proposed Updates
+
+### WorkflowInterceptor.InterceptWorkflow
+
+Current:
+
+```go
+InterceptWorkflow(info *WorkflowInfo, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+```
+
+Proposal:
+
+```go
+InterceptWorkflow(ctx Context, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+```
+
+Why: Callers can use the context to get workflow info.
+
+### WorkflowXCallsInterceptor
+
+Proposal: Change the name to `WorkflowXInterceptor`. Why: `Calls` just adds noise (more accepted in Java, not Go).
+
+### WorkflowInboundCallsInterceptor.ExecuteWorkflow
+
+Current:
+
+```go
+ExecuteWorkflow(ctx Context, workflowType string, args ...interface{}) []interface{}
+```
+
+Proposal:
+
+```go
+type WorkflowInput struct {
+  Args []interface{}
+}
+
+ExecuteWorkflow(ctx Context, input *WorkflowInput) (interface{}, error)
```

Comment:
Should this be called `Execute`?

Reply:
I don't think so for a couple of reasons. Firstly, in Go it's usually <verb><noun> unless the noun is the subject of the method (i.e. if this were a workflow instead of a workflow interceptor, `Execute` makes more sense, otherwise you're not "executing" the interceptor, you're intercepting the execution). Also this matches the name used in the client which I think makes it clearer. However, if we do want to be consistent with signal and query, I could rename this to `HandleExecution`.

Reply:
I don't know what the best practices in go are, I'm fine with whatever we choose here.
Let's make sure the names are consistent though.